### PR TITLE
Remove marking the extension as Failed

### PR DESCRIPTION
**Pull request by pgombar**
## Description

Today, during an extension upgrade, if the update command of the new extension fails, the new extension will be marked as Failed. This is the only usage of that extension handler state.

However, this means that, when a new goal state comes in, since the new extension is marked as Failed, agent will try to call enable on the new extension, skipping update, because the extension handler's state is not equal to [NotInstalled](https://github.com/Azure/WALinuxAgent/blob/develop/azurelinuxagent/ga/exthandlers.py#L487).

This PR ensures that the entire upgrade scenario is called even if the previous goal state resulted in a failed update step. This is done by removing the Failed state as an option and, as a result, when the update command fails, the new extension will be removed, so when the new goal state comes in, it will be treated as a new - not previously installed - extension and the entire upgrade scenario will be exercised.

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1726)

### CODE REVIEW

File: azurelinuxagent/ga/exthandlers.py [hunk 1]
```diff
@@ -1023,19 +1022,14 @@ def update(self, version=None, disable_exit_code=None, updating_from_version=Non
...
-            self.set_handler_state(ExtHandlerState.Failed)
-            raise
+        self.set_operation(WALAEventOperation.Update)
+        man = self.load_manifest()
+        update_cmd = man.get_update_command()
+        self.logger.info("Update extension [{0}]".format(update_cmd))
+        self.launch_command(update_cmd,
+                            timeout=900,
+                            extension_error_code=ExtensionErrorCodes.PluginUpdateProcessingFailed,
+                            env=env)
```
Comment (narrieta) [inline azurelinuxagent/ga/exthandlers.py]: let's double-check report_ext_handler_status --- seems like the update failure would have been reported to crp

Comment (larohra) [reply azurelinuxagent/ga/exthandlers.py]: Also double check the PR where they added the FAILED state, they did it to deal with going in the infinite loop I think. The PR would give you the motivation behind adding that and to see if this doesnt reintroduce the problem

File: tests/ga/test_extension.py [hunk 2]
```diff
@@ -1464,24 +1464,32 @@ def test_uninstall_failure(self, patch_get_uninstall_command, *args):
         self._assert_no_handler_status(protocol.report_vm_status)
 
     @patch('azurelinuxagent.ga.exthandlers.HandlerManifest.get_update_command')
-    def test_upgrade_failure(self, patch_get_update_command, *args):
+    def test_extension_upgrade_failure_when_update_fails(self, patch_get_update_command, *args):
         """
-        Extension upgrade failure should not be retried
+        When the update command of the new extension fails, it should result in the new extension removed and the
+        old extension disabled. On the next goal state, the entire upgrade scenario should be retried (once),
+        meaning the new extension is downloaded again and update is called again.
```
Comment (larohra) [inline tests/ga/test_extension.py]: If there's some bug in the update.cmd of the extension, wont this cause the agent to download and delete the zip file at every GS? Aesthetically that would make sense since the code is buggy, but it would still cause extra network bandwidth for the customer. Should we handle this scenario?

Comment (narrieta) [reply tests/ga/test_extension.py]: Initially I was thinking along those lines, too. But since now we go in the loop only on new goal state, my thought was that re-downloading was ok. What do you guys think? @pgombar @vrdmr

File: tests/ga/test_extension.py [hunk 3]
```diff
@@ -1464,24 +1464,32 @@ def test_uninstall_failure(self, patch_get_uninstall_command, *args):
...
         """
         test_data, exthandlers_handler, protocol = self._set_up_update_test_and_update_gs(patch_get_update_command,
                                                                                           *args)
 
-        exthandlers_handler.run()
-        self.assertEqual(1, patch_get_update_command.call_count)
+        with patch('azurelinuxagent.ga.exthandlers.HandlerManifest.get_enable_command') as patch_get_enable_command:
+            exthandlers_handler.run()
+            self.assertEqual(1, patch_get_update_command.call_count)
```
Comment (narrieta) [inline tests/ga/test_extension.py]: I've always felt that those asserts on the get_*_command functions are too weak. As the code evolves they may be other reason why they are not called. For future tests maybe we can think of stronger asserts? (maybe mock popen and assert that the corresponding command is never invoked?)

Comment (narrieta) [reply tests/ga/test_extension.py]: for example, below, at line 1483... the getter may have been called, but (as the code evolves) maybe the command was never executed

File: tests/ga/test_extension.py (hunk 2 above)
Comment (larohra) [reply tests/ga/test_extension.py]: My concern is mostly about those extensions that the customer doesnt even want or know about (like OMS) and which are extremely buggy. Extensions like this would cause unnecessary traffic which might make them mad. Not sure if that is still part of the scope of this PR

File: tests/ga/test_extension.py [hunk 4]
```diff
@@ -1464,24 +1464,32 @@ def test_uninstall_failure(self, patch_get_uninstall_command, *args):
...
-        # On the next iteration, update should not be retried
-        exthandlers_handler.run()
-        self.assertEqual(1, patch_get_update_command.call_count)
+            # On the next iteration, update should be retried.
+            exthandlers_handler.run()
+            self.assertEqual(1, patch_get_update_command.call_count)
+            self.assertEqual(0, patch_get_enable_command.call_count)
 
-        self._assert_handler_status(protocol.report_vm_status, "NotReady", expected_ext_count=1, version="1.0.1")
+        self.assertEquals("Ready", protocol.report_vm_status.call_args[0][0].vmAgent.status)
```
Comment (narrieta) [inline tests/ga/test_extension.py]: why the change from NotReady to Ready?

File: tests/ga/test_extension.py [hunk 5]
```diff
@@ -1464,24 +1464,32 @@ def test_uninstall_failure(self, patch_get_uninstall_command, *args):
...
+
+        # Since the old extension is disabled, and the new one never got enabled, there is nothing to report.
+        self._assert_no_handler_status(protocol.report_vm_status)
```
Comment (narrieta) [inline tests/ga/test_extension.py]: as discussed offline, the failure updating to the new version should be reported.

File: tests/ga/test_extension.py (hunk 2 above)
Comment (vrdmr) [reply tests/ga/test_extension.py]: > wont this cause the agent to download and delete the zip file at every GS?

The zip files for the new update would still be downloaded but only one-time per GS (not infinite). I feel it is okay. Yes, our downloads might increase in a fast-changing VM, but still would be far less than the earlier problem (similar to what @narrieta said)

The bug of infinite that came back in 2.2.44 was due to the issue that cleanup code does not have the ExtHandler models of the updated extension and thus assumes it is an outdated version. And there is no easy way to fix that problem if we still want to allow infinite attempts.

File: tests/ga/test_extension.py [hunk 6]
```diff
@@ -1464,24 +1464,32 @@ def test_uninstall_failure(self, patch_get_uninstall_command, *args):
...
         """
         test_data, exthandlers_handler, protocol = self._set_up_update_test_and_update_gs(patch_get_update_command,
                                                                                           *args)
 
-        exthandlers_handler.run()
-        self.assertEqual(1, patch_get_update_command.call_count)
+        with patch('azurelinuxagent.ga.exthandlers.HandlerManifest.get_enable_command') as patch_get_enable_command:
+            exthandlers_handler.run()
+            self.assertEqual(1, patch_get_update_command.call_count)
+            self.assertEqual(0, patch_get_enable_command.call_count)
```
Comment (vrdmr) [inline tests/ga/test_extension.py]: Can you also assert if the error code returned was `ExtensionErrorCodes.PluginUpdateProcessingFailed` to make the assert stronger?

File: tests/ga/test_extension.py (hunk 4 above)
Comment (vrdmr) [inline tests/ga/test_extension.py]: Why is it `Ready`? If the Update failed?

File: tests/ga/test_extension.py (hunk 2 above)
Comment (pgombar) [reply tests/ga/test_extension.py]: I think it's ok to have the extension get downloaded once per goal state since we got rid of the upgrade loop.

File: tests/ga/test_extension.py (hunk 4 above)
Comment (pgombar) [reply tests/ga/test_extension.py]: You are comparing two different things -- the old NotReady assertion was testing the handler status, this assertion is testing the agent status. Testing the agent status isn't really needed here, I added it as an extra precaution. I will update the tests to make it clearer.

Comment (pgombar) [reply tests/ga/test_extension.py]: See my comment above. I will send an update to make it clearer.

File: tests/ga/test_extension.py (hunk 5 above)
Comment (pgombar) [reply tests/ga/test_extension.py]: Updating the tests to reflect this assertion earlier, after the first `exthandlers_handler.run()`, when the failure is reported.

File: tests/ga/test_extension.py (hunk 4 above)
Comment (pgombar) [reply tests/ga/test_extension.py]: Just in case the ordering of the comments changes, I'm referring to this reply:

You are comparing two different things -- the old NotReady assertion was testing the handler status, this assertion is testing the agent status. Testing the agent status isn't really needed here, I added it as an extra precaution. I will update the tests to make it clearer.

**Comment by pgombar** [comment]
Abandoning in favour of #1731.
